### PR TITLE
add  clb targetgroup

### PR DESCRIPTION
**Pull request by ChuGuai**
terraform-provider-tencentcloud guojunchu$ make testacc TESTARGS="-run TestAccTencentCloudClbTGAttachmentInstance_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run TestAccTencentCloudClbTGAttachmentInstance_basic -timeout 120m
?       github.com/tencentcloudstack/terraform-provider-tencentcloud    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc     (cached) [no tests to run]
=== RUN   TestAccTencentCloudClbTGAttachmentInstance_basic
--- PASS: TestAccTencentCloudClbTGAttachmentInstance_basic (13.08s)
PASS
ok      github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud       16.360s
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity  [no test files]
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper       [no test files]
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit     [no test files]

terraform-provider-tencentcloud guojunchu$ make testacc TESTARGS="-run TestAccTencentCloudClbTargetGroup_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run TestAccTencentCloudClbTargetGroup_basic -timeout 120m
?       github.com/tencentcloudstack/terraform-provider-tencentcloud    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc     (cached) [no tests to run]
=== RUN   TestAccTencentCloudClbTargetGroup_basic
--- PASS: TestAccTencentCloudClbTargetGroup_basic (7.27s)
PASS
ok      github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud       8.845s
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity  [no test files]
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper       [no test files]
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit     [no test files]

terraform-provider-tencentcloud guojunchu$ make testacc TESTARGS="-run TestAccTencentClbTargetGroupAttachmentResource"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run TestAccTencentClbTargetGroupAttachmentResource -timeout 120m
?       github.com/tencentcloudstack/terraform-provider-tencentcloud    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc     0.971s [no tests to run]
=== RUN   TestAccTencentClbTargetGroupAttachmentResource
--- PASS: TestAccTencentClbTargetGroupAttachmentResource (29.81s)
PASS
ok      github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud       30.979s
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity  [no test files]
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper       [no test files]
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit     [no test files]

terraform-provider-tencentcloud guojunchu$ make testacc TESTARGS="-run TestAccTencentCloudDataSourceClbTargetGroup"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run TestAccTencentCloudDataSourceClbTargetGroup -timeout 120m
?       github.com/tencentcloudstack/terraform-provider-tencentcloud    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc     0.555s [no tests to run]
=== RUN   TestAccTencentCloudDataSourceClbTargetGroup
--- PASS: TestAccTencentCloudDataSourceClbTargetGroup (40.37s)
PASS
ok      github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud       41.548s
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity  [no test files]
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper       [no test files]
?       github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit     [no test files]

### CODE REVIEW

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
+/*
+Provides a resource to create a CLB target group instance attachment.
+
+Example Usage
+
+```hcl
+resource "tencentcloud_clb_tg_instance_attachment" "test"{
+    target_group_id = "lbtg-3k3io0i0"
```
Comment (brickzzhang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:8]: 格式化

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+				Type:        schema.TypeInt,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "A new port for the target group instance.",
+			},
+		},
+	}
+}
+func resourceTencentCloudClbTGAttachmentInstanceCreate(d *schema.ResourceData, meta interface{}) error {
+	var (
```
Comment (brickzzhang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:77]: defer logelapsed

File: tencentcloud/resource_tc_clb_tg_instance_attachment_test.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+				return nil
+			}
+		}
+		return fmt.Errorf("[TECENT_TERRAFORM_CHECK][CLB targetGroupinstance attachment][Exists] id %s is not exist", rs.Primary.ID)
+	}
+}
+
+const testAccClbTGAttachmentInstance_basic = `
+resource "tencentcloud_clb_tg_instance_attachment" "test"{
+    target_group_id = "lbtg-3k3io0i0"
```
Comment (brickzzhang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment_test.go:114]: 格式化

File: tencentcloud/service_tencentcloud_clb.go [hunk 4]
```diff
@@ -890,3 +890,304 @@ func clbNewTarget(instanceId, port, weight interface{}) *clb.Target {
...
+			logId, request.GetAction(), request.ToJsonString(), response.ToJsonString())
+		return nil
+	})
+
+	errRet = err
+	return
+}
+
+func isExistCode(code string) bool {
+	codeMap := map[string]bool{
```
Comment (brickzzhang) [inline tencentcloud/service_tencentcloud_clb.go:1182]: 放到extension中吧

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,4 +1,16 @@
-## 1.44.1 (Unreleased)
+## 1.45.0 (Unreleased)
```
Comment (gailwang) [inline CHANGELOG.md:1]: 增加小特性 使用1.44.1

File: tencentcloud/provider.go [hunk 6]
```diff
@@ -709,6 +714,9 @@ func Provider() terraform.ResourceProvider {
 			"tencentcloud_clb_listener_rule":               resourceTencentCloudClbListenerRule(),
 			"tencentcloud_clb_attachment":                  resourceTencentCloudClbServerAttachment(),
 			"tencentcloud_clb_redirection":                 resourceTencentCloudClbRedirection(),
+			"tencentcloud_clb_targetgroup":                 resourceTencentCloudClbTargetGroup(),
```
Comment (gailwang) [inline tencentcloud/provider.go:717]: targetgroup target_group tg 建议统一一种

File: tencentcloud/resource_tc_clb_listener_rule_test.go [hunk 7]
```diff
@@ -27,6 +27,20 @@ func TestAccTencentCloudClbListenerRule_basic(t *testing.T) {
...
+				Config: testAccClbListenerRule_basic_update,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckClbListenerRuleExists("tencentcloud_clb_listener_rule.rule_basic"),
+					resource.TestCheckResourceAttrSet("tencentcloud_clb_listener_rule.rule_basic", "clb_id"),
+					resource.TestCheckResourceAttrSet("tencentcloud_clb_listener_rule.rule_basic", "listener_id"),
+					resource.TestCheckResourceAttr("tencentcloud_clb_listener_rule.rule_basic", "domain", "abc.com"),
+					resource.TestCheckResourceAttr("tencentcloud_clb_listener_rule.rule_basic", "session_expire_time", "30"),
+					resource.TestCheckResourceAttr("tencentcloud_clb_listener_rule.rule_basic", "url", "/"),
+					resource.TestCheckResourceAttr("tencentcloud_clb_listener_rule.rule_basic", "scheduler", "WRR"),
+					resource.TestCheckResourceAttr("tencentcloud_clb_listener_rule.rule_basic", "target_type", "NODE"),
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_listener_rule_test.go:43]: 如果这个target_type是forceNew的话，直接就destroy and create了，update的校验其实没有必要，只要校验这个create的就可以了

File: tencentcloud/resource_tc_clb_target_group_attachment.go [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+		logId       = getLogId(contextNil)
+		ctx         = context.WithValue(context.TODO(), logIdKey, logId)
+		id          = d.Id()
+		targetInfos []*clb.TargetGroupInfo
+		err         error
+	)
+
+	ids := strings.Split(id, FILED_SP)
+	if len(ids) < 4 {
+		return fmt.Errorf("clb target group attachment id must contains clb_id, listernrt_id, targrt_group_id, rule_id")
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_attachment.go:253]: clb CLB统一一下

File: tencentcloud/resource_tc_clb_target_group_attachment_test.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+					var originLocationId string
+					originClbId := *rule.LoadBalancerId
+					originListenerId := *rule.ListenerId
+					if rule.LocationId != nil {
+						originLocationId = *rule.LocationId
+					}
+
+					if originListenerId == ids[1] && originClbId == ids[2] && originLocationId == ids[3] {
+						return resource.RetryableError(
+							fmt.Errorf("Rule association target group instance still exist. [targetGroupId=%s, listenerId=%s, cldId=%s, ruleId=%s].",
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_attachment_test.go:80]: 测试是否destory的时候，并不需要给retry，只检查一次。

File: tencentcloud/resource_tc_clb_targetgroup.go [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Target group name.",
+			},
+			"vpc_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "0",
+				ForceNew:    true,
+				Description: "VPC ID, default is Based on the network.",
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_targetgroup.go:53]: 注意大小写 Based

File: tencentcloud/resource_tc_clb_targetgroup.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+		vpcId           string
+		port            int
+		targetGroupId   string
+		inErr, outErr   error
+	)
+
+	if temp, ok := d.GetOk("target_group_name"); ok {
+		targetGroupName = temp.(string)
+	}
+	if temp, ok := d.GetOk("vpc_id"); ok {
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_targetgroup.go:89]: 有设置default的时候可以直接取

File: tencentcloud/resource_tc_clb_targetgroup.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+		insAttachments  = make([]*clb.TargetGroupInstance, 0)
+		targetGroupName string
+		vpcId           string
+		port            int
+		targetGroupId   string
+		inErr, outErr   error
+	)
+
+	if temp, ok := d.GetOk("target_group_name"); ok {
+		targetGroupName = temp.(string)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_targetgroup.go:87]: 这个参数不设置的情况测试过吗？

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+		bindIp        = d.Get("bind_ip").(string)
+		port          = d.Get("port").(int)
+		weight        int
+		outErr, inErr error
+	)
+	if temp, ok := d.GetOk("weight"); ok {
+		weight = temp.(int)
+	}
+
+	outErr = resource.Retry(readRetryTimeout, func() *resource.RetryError {
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:95]: create 使用writeRetryTimeout

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+		inErr = clbService.RegisterTargetInstances(ctx, targetGroupId, bindIp, uint64(port), uint64(weight))
+		if inErr != nil {
+			return retryError(inErr)
+		}
+		return nil
+	})
+	if outErr != nil {
+		log.Printf("[CRITAL]%s read CLB targetgroup instance attachment failed, reason:%+v", logId, outErr)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:103]: create log有错误

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+		return outErr
+	}
+	time.Sleep(time.Duration(3) * time.Second)
+
+	d.SetId(targetGroupId + FILED_SP + bindIp + FILED_SP + strconv.Itoa(port) + FILED_SP + strconv.Itoa(weight))
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:108]: 这里是否targetGroupId和Ip和port不足以确定这一个绑定资源呢？同一个绑定可以多次使用不同权重得到两个资源吗

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+		clbService = ClbService{client: meta.(*TencentCloudClient).apiV3Conn}
+		id         = d.Id()
+	)
+	idSplit := strings.Split(id, FILED_SP)
+	if len(idSplit) < 4 {
+		return fmt.Errorf("targetgroup instance attachment id is not set")
+	}
+	targetGroupId := idSplit[0]
+	bindIp := idSplit[1]
+	port, _ := strconv.ParseUint(idSplit[2], 0, 64)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:128]: err需要写出来
port, err := strconv.ParseUint(idSplit[2], 0, 64)
if err != nil{
  return err
}

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+	weight, _ := strconv.ParseUint(idSplit[3], 0, 64)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:129]: 同上，不可忽略

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+	filters := make(map[string]string)
+	filters["TargetGroupId"] = targetGroupId
+	filters["BindIP"] = bindIp
+	targetGroupInstances, err := clbService.DescribeTargetGroupInstances(ctx, filters)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:134]: 这里没有retry

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+	for _, tgInstance := range targetGroupInstances {
+		if *tgInstance.Weight == weight && *tgInstance.Port == port {
+			_ = d.Set("target_group_id", idSplit[0])
+			_ = d.Set("bind_ip", idSplit[1])
+			_ = d.Set("port", idSplit[2])
+			_ = d.Set("weight", idSplit[3])
+			return nil
+		}
+	}
+	return fmt.Errorf("there is no targetGroupInstance attachment by id %s", id)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:147]: 没有读到直接SetId("") 不要报错

File: tencentcloud/resource_tc_clb_tg_instance_attachment.go [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+
+	var (
+		logId      = getLogId(contextNil)
+		ctx        = context.WithValue(context.TODO(), logIdKey, logId)
+		clbService = ClbService{client: meta.(*TencentCloudClient).apiV3Conn}
+		id         = d.Id()
+	)
+	idSplit := strings.Split(id, FILED_SP)
+	if len(idSplit) < 4 {
+		d.SetId("")
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_tg_instance_attachment.go:161]: Id有问题不要SetId("")
报错出来

File: tencentcloud/service_tencentcloud_clb.go [hunk 21]
```diff
@@ -890,3 +890,304 @@ func clbNewTarget(instanceId, port, weight interface{}) *clb.Target {
...
+	for k, v := range filters {
+		filter := clb.Filter{
+			Name:   helper.String(k),
+			Values: []*string{helper.String(v)},
+		}
+		request.Filters = append(request.Filters, &filter)
+	}
+
+	var offset uint64 = 0
+	var pageSize uint64 = 100
```
Comment (gailwang) [inline tencentcloud/service_tencentcloud_clb.go:993]: 常量写在extension内，clb有提供这个CLB_PAGE_LIMIT，可以直接使用

File: tencentcloud/service_tencentcloud_clb.go [hunk 22]
```diff
@@ -890,3 +890,304 @@ func clbNewTarget(instanceId, port, weight interface{}) *clb.Target {
...
```
Comment (gailwang) [inline tencentcloud/service_tencentcloud_clb.go:1033]: pageSize := int64(CLB_PAGE_LIMIT)

File: tencentcloud/service_tencentcloud_clb.go [hunk 23]
```diff
@@ -890,3 +890,304 @@ func clbNewTarget(instanceId, port, weight interface{}) *clb.Target {
...
+		},
+	}
+
+	err := resource.Retry(writeRetryTimeout, func() *resource.RetryError {
+		ratelimit.Check(request.GetAction())
+		response, err := me.client.UseClbClient().DisassociateTargetGroups(request)
+		if err != nil {
+			if e, ok := err.(*sdkErrors.TencentCloudSDKError); ok {
+				code := e.GetCode()
+				if isExistCode(code) {
```
Comment (gailwang) [inline tencentcloud/service_tencentcloud_clb.go:1163]: 这里如果确认retryError()函数不覆盖的情况，再这么写

File: tencentcloud/resource_tc_clb_listener_rule.go [hunk 24]
```diff
@@ -204,6 +212,9 @@ func resourceTencentCloudClbListenerRuleCreate(d *schema.ResourceData, meta inte
 	rule.Domain = helper.String(domain)
 	url := d.Get("url").(string)
 	rule.Url = helper.String(url)
+	if v, ok := d.GetOk("target_type"); ok {
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_listener_rule.go:215]: 有default值可以直接取

File: tencentcloud/resource_tc_clb_target_group.go [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+
+	outErr = resource.Retry(writeRetryTimeout, func() *resource.RetryError {
+		targetGroupId, inErr = clbService.CreateTargetGroup(ctx, targetGroupName, vpcId, uint64(port), insAttachments)
+		if inErr != nil {
+			return retryError(inErr)
+		}
+		return nil
+	})
+	if outErr != nil {
+		log.Printf("[CRITAL]%s read CLB targetgroup failed, reason:%+v", logId, outErr)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group.go:99]: 这个是create的error log信息有误

File: tencentcloud/resource_tc_clb_target_group.go [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+		clbService = ClbService{client: meta.(*TencentCloudClient).apiV3Conn}
+		id         = d.Id()
+	)
+	filters := make(map[string]string)
+	targetGroupInfos, err := clbService.DescribeTargetGroups(ctx, id, filters)
+	if err != nil {
+		return err
+	}
+	if len(targetGroupInfos) < 1 {
+		return fmt.Errorf("There is no clb targetGroup, targetGroupId : %s", id)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group.go:124]: Read内没有找到不要报错

Comment (gailwang) [reply tencentcloud/resource_tc_clb_target_group.go:124]: clb CLB统一

File: tencentcloud/resource_tc_clb_target_group.go [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+	var (
+		logId         = getLogId(contextNil)
+		ctx           = context.WithValue(context.TODO(), logIdKey, logId)
+		clbService    = ClbService{client: meta.(*TencentCloudClient).apiV3Conn}
+		targetGroupId = d.Id()
+	)
+
+	if d.HasChange("target_group_name") || d.HasChange("port") {
+		targetGroupName := d.Get("target_group_name").(string)
+		port := d.Get("port").(int)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group.go:145]: 这里本来port是optional 如果从设置变为没有设置，d.Get可能读不到。看下这里port是否有default值。

File: tencentcloud/resource_tc_clb_target_group_attachment.go [hunk 28]
```diff
@@ -0,0 +1,290 @@
...
+	})
+	if err != nil {
+		log.Printf("[CRITAL]%s read CLB instance failed, reason:%+v", logId, err)
+		return err
+	}
+	if instance == nil {
+		return fmt.Errorf("CLB instance not exist")
+	}
+
+	//check listener_id
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_attachment.go:136]: 如果直接测试不存在的clb clb listener 去绑定targetGroup SDK应该会直接报错，这里的校验应该不需要

File: tencentcloud/resource_tc_clb_target_group_attachment.go [hunk 29]
```diff
@@ -0,0 +1,290 @@
...
+	if len(ids) < 4 {
+		return fmt.Errorf("clb target group attachment id must contains clb_id, listernrt_id, targrt_group_id, rule_id")
+	}
+
+	err := clbService.DescribeAssociateTargetGroups(ctx, ids)
+	if err != nil {
+		return err
+	}
+
+	errs := []error{
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_attachment.go:224]: 这里的errs 写法不需要，直接d.Set

File: tencentcloud/resource_tc_clb_target_group_attachment.go [hunk 30]
```diff
@@ -0,0 +1,290 @@
...
+		clbService = ClbService{
+			client: meta.(*TencentCloudClient).apiV3Conn,
+		}
+		logId = getLogId(contextNil)
+		ctx   = context.WithValue(context.TODO(), logIdKey, logId)
+		id    = d.Id()
+	)
+
+	ids := strings.Split(id, FILED_SP)
+	if len(ids) < 4 {
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_attachment.go:215]: != 4 这里是代码里定下的规则，>4的情况也属于非法

File: tencentcloud/resource_tc_clb_target_group_attachment.go [hunk 31]
```diff
@@ -0,0 +1,290 @@
...
+		return err
+	}
+
+	errs := []error{
+		d.Set("targrt_group_id", ids[0]),
+		d.Set("listener_id", ids[1]),
+		d.Set("clb_id", ids[2]),
+		d.Set("rule_id", ids[3]),
+	}
+	if len(errs) > 0 {
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_attachment.go:230]: 不需要，参考其他resource的赋值

File: tencentcloud/resource_tc_clb_target_group_attachment.go [hunk 32]
```diff
@@ -0,0 +1,290 @@
...
+		targetInfos []*clb.TargetGroupInfo
+		err         error
+	)
+
+	ids := strings.Split(id, FILED_SP)
+	if len(ids) < 4 {
+		return fmt.Errorf("clb target group attachment id must contains clb_id, listernrt_id, targrt_group_id, rule_id")
+	}
+
+	if err := clbService.DisassociateTargetGroups(ctx, ids[0], ids[1], ids[2], ids[3]); err != nil {
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_attachment.go:256]: 这里没有重试

File: tencentcloud/resource_tc_clb_target_group_attachment_test.go [hunk 33]
```diff
@@ -0,0 +1,159 @@
...
+		}
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("clb target group attachment id is not set")
+		}
+		clbService := ClbService{
+			client: testAccProvider.Meta().(*TencentCloudClient).apiV3Conn,
+		}
+
+		ids := strings.Split(rs.Primary.ID, FILED_SP)
+		if len(ids) < 4 {
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_attachment_test.go:114]: 校验Id的时候按照自己定的规则来

File: tencentcloud/resource_tc_clb_target_group_instance_attachment.go [hunk 34]
```diff
@@ -0,0 +1,264 @@
...
+			"port": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Port of the target group instance.",
+			},
+			"weight": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				ForceNew:    true,
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_instance_attachment.go:135]: https://cloud.tencent.com/document/product/214/40550
权重可以通过这个接口修改，不要去forceNew

File: tencentcloud/resource_tc_clb_target_group_instance_attachment.go [hunk 35]
```diff
@@ -0,0 +1,264 @@
...
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateNotEmpty,
+				ForceNew:     true,
+				Description:  "The Intranet IP of the target group instance.",
+			},
+			"port": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_instance_attachment.go:129]: https://cloud.tencent.com/document/product/214/40551
端口可以通过这个接口修改，不要去forceNew

File: tencentcloud/service_tencentcloud_clb.go [hunk 36]
```diff
@@ -890,3 +890,278 @@ func clbNewTarget(instanceId, port, weight interface{}) *clb.Target {
...
+				originLocationId = *rule.LocationId
+			}
+
+			if originListenerId == ids[1] && originClbId == ids[2] && originLocationId == ids[3] {
+				return nil
+			}
+		}
+	}
+
+	return fmt.Errorf("Rule association target group instance does not exist. [targetGroupId=%s, listenerId=%s, cldId=%s, ruleId=%s].",
```
Comment (gailwang) [inline tencentcloud/service_tencentcloud_clb.go:1129]: Describe不要返回没有找到的错误

Comment (gailwang) [reply tencentcloud/service_tencentcloud_clb.go:1129]: 使用 返回  has, error的返回，找到了赋值has = true
没有找到直接return

File: tencentcloud/service_tencentcloud_clb.go [hunk 37]
```diff
@@ -890,3 +890,278 @@ func clbNewTarget(instanceId, port, weight interface{}) *clb.Target {
...
+	}
+
+	err := resource.Retry(writeRetryTimeout, func() *resource.RetryError {
+		ratelimit.Check(request.GetAction())
+		response, err := me.client.UseClbClient().DisassociateTargetGroups(request)
+		if err != nil {
+			return retryError(err, InternalError)
+		}
+		if response == nil {
+			return resource.RetryableError(fmt.Errorf("DisassociateTargetGroups response is nil"))
```
Comment (gailwang) [inline tencentcloud/service_tencentcloud_clb.go:1158]: 空返回不是可以retryable的，直接Unretryable

File: tencentcloud/service_tencentcloud_clb.go [hunk 38]
```diff
@@ -890,3 +890,278 @@ func clbNewTarget(instanceId, port, weight interface{}) *clb.Target {
...
+	}
+
+	err := resource.Retry(writeRetryTimeout, func() *resource.RetryError {
+		ratelimit.Check(request.GetAction())
+		response, err := me.client.UseClbClient().AssociateTargetGroups(request)
+		if err != nil {
+			return retryError(err, InternalError)
+		}
+		if response == nil {
+			return resource.RetryableError(fmt.Errorf("AssociateTargetGroups response is nil"))
```
Comment (gailwang) [inline tencentcloud/service_tencentcloud_clb.go:1080]: 空返回不是Retryable

File: tencentcloud/resource_tc_clb_targetgroup.go [hunk 39]
```diff
@@ -0,0 +1,181 @@
...
+		clbService = ClbService{client: meta.(*TencentCloudClient).apiV3Conn}
+		id         = d.Id()
+	)
+	filters := make(map[string]string)
+	targetGroupInfos, err := clbService.DescribeTargetGroups(ctx, id, filters)
+	if err != nil {
+		return err
+	}
+	if len(targetGroupInfos) < 1 {
+		return fmt.Errorf("There is no clb targetGroup, targetGroupId : %s", id)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_targetgroup.go]: read不到的时候不要报错

File: tencentcloud/resource_tc_clb_target_group_attachment.go (hunk 32 above)
Comment (ttomzhou) [reply tencentcloud/resource_tc_clb_target_group_attachment.go:237]: 在service层里面有重试逻辑

File: tencentcloud/resource_tc_clb_target_group_attachment.go (hunk 28 above)
Comment (ttomzhou) [reply tencentcloud/resource_tc_clb_target_group_attachment.go]: 保留了校验rule的后端目标组，复制传入的rule里面没有开启后端目标组，SDK会报FailedOperation错误，会导致无用的重试

File: tencentcloud/resource_tc_clb_target_group_attachment_test.go [hunk 40]
```diff
@@ -0,0 +1,159 @@
...
+			continue
+		}
+
+		ids := strings.Split(rs.Primary.ID, FILED_SP)
+		if len(ids) != 4 {
+			return fmt.Errorf("CLB target group attachment id must contains clb_id, listernrt_id, targrt_group_id, rule_id")
+		}
+
+		err = resource.Retry(readRetryTimeout, func() *resource.RetryError {
+			targetInfos, err = clbService.DescribeTargetGroups(ctx, ids[0], nil)
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_attachment_test.go:63]: test检查类型的函数不需要重试

**Review by gailwang — APPROVED** [review]

**Review by crab21 — APPROVED** [review]

File: tencentcloud/resource_tc_clb_target_group_instance_attachment.go [hunk 41]
```diff
@@ -0,0 +1,285 @@
...
+		logId         = getLogId(contextNil)
+		ctx           = context.WithValue(context.TODO(), logIdKey, logId)
+		clbService    = ClbService{client: meta.(*TencentCloudClient).apiV3Conn}
+		targetGroupId = d.Get("target_group_id").(string)
+		bindIp        = d.Get("bind_ip").(string)
+		port          = d.Get("port").(int)
+		weight        int
+		err           error
+	)
+	if temp, ok := d.GetOk("weight"); ok {
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_instance_attachment.go:154]: requried字段不需要用GetOk

File: tencentcloud/resource_tc_clb_target_group_instance_attachment.go [hunk 42]
```diff
@@ -0,0 +1,285 @@
...
+	targetGroupId = idSplit[0]
+	bindIp = idSplit[1]
+	port, err = strconv.Atoi(idSplit[2])
+	if err != nil {
+		return err
+	}
+
+	if d.HasChange("weight") {
+		newWeight := d.Get("weight").(int)
+		err := clbService.ModifyTargetGroupInstancesWeight(ctx, targetGroupId, bindIp, uint64(port), uint64(newWeight))
```
Comment (gailwang) [inline tencentcloud/resource_tc_clb_target_group_instance_attachment.go:246]: 这里漏了重试逻辑

File: tencentcloud/resource_tc_clb_target_group.go [hunk 43]
```diff
@@ -0,0 +1,160 @@
...
+			},
+			"vpc_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "0",
+				ForceNew:    true,
+				Description: "VPC ID, default is based on the network.",
+			},
+			//computed
+			"target_group_id": {
```
Comment (oliverpei) [inline tencentcloud/resource_tc_clb_target_group.go:55]: 这个不需要，默认有个id的属性

File: tencentcloud/resource_tc_clb_target_group_attachment.go [hunk 44]
```diff
@@ -0,0 +1,270 @@
...
+	ids := strings.Split(id, FILED_SP)
+	if len(ids) != 4 {
+		return fmt.Errorf("CLB target group attachment id must contains clb_id, listernrt_id, targrt_group_id, rule_id")
+	}
+
+	has, err := clbService.DescribeAssociateTargetGroups(ctx, ids)
+	if err != nil {
+		return err
+	}
+	if !has {
```
Comment (oliverpei) [inline tencentcloud/resource_tc_clb_target_group_attachment.go:206]: d.SetId("")

File: tencentcloud/resource_tc_clb_target_group_attachment.go [hunk 45]
```diff
@@ -0,0 +1,270 @@
...
+			return retryError(err, InternalError)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	for _, rule := range listener.Rules {
+		if locationId == *rule.LocationId && (rule.TargetType != nil && *rule.TargetType == CLB_TARGET_TYPE_TARGETGROUP) {
+			isRuleExist = true
```
Comment (Himer) [inline tencentcloud/resource_tc_clb_target_group_attachment.go:131]: break

File: tencentcloud/resource_tc_clb_target_group.go [hunk 46]
```diff
@@ -0,0 +1,160 @@
...
+	targetGroupInfos, err := clbService.DescribeTargetGroups(ctx, id, filters)
+	if err != nil {
+		return err
+	}
+	if len(targetGroupInfos) < 1 {
+		d.SetId("")
+		return nil
+	}
+	_ = d.Set("target_group_name", targetGroupInfos[0].TargetGroupName)
+	_ = d.Set("vpc_id", targetGroupInfos[0].VpcId)
```
Comment (Himer) [inline tencentcloud/resource_tc_clb_target_group.go:101]: default vpc 可能导致forcenew

**Review by gailwang — APPROVED** [review]

**Pull request merged**